### ISSUE TITLE
Paypal Transaction Security/Verification Post Processing

### ISSUE THREAD

**Issue by mooror** [i-lateral/silverstripe-checkout#6]
Hey there, I was reading through the code in your repository and I was not able to find any verification functionality for double checking that the checkout form was not tampered with before it was sent. I see that you have created an IPN listener to protect against request forgery. But I cannot seem to find any code in your repo that compares the returned POST variables against DB records to ensure that the item information matches. 

It is possible that I missed it when I read through your code. But if on the other hand, you do not ever check the returned POST request variables you really need to. Otherwise, a malicious user could edit the hidden fields in the purchase form to change the price of an item before submitting it. And because they are editing the form before it reaches PayPal (and not trying to forge a Paypal request to send directly to your server) they will subvert any basic IPN checks. 

Please get back to me as soon as possible. I happen to use this module on a production site and would like to either a) patch it as quickly as possible if there is a security hole or b) stop worrying about it if you already have functionality that protects against this kind of attack.

Thanks in advance,
Mooror

**Comment by mlewis-everley** [comment]
Hmm, the answer should depend on what branch you are using.

If you are on the `master` branch, then this should be handled via: https://github.com/i-lateral/silverstripe-checkout/blob/master/code/control/payment/PayPalHandler.php#L183

If you are using omnipay (`3.0`), I _believe_ that omnipay handles this (but I cannot remember off the top of my head).

Cheers,

Mo

**Comment by mooror** [comment]
Yeah, I read through that part of your code before posting this. As far as I can see you are using the default `PaypalIPN` class that Paypal provides to check if the IPN was verified. If it was verified then you check and store the status and continue with the request. 

This is is a good and necessary step, but extra checks must be made after this. Paypal's [IPN Protocol and Architecture Documentation](https://developer.paypal.com/docs/classic/ipn/integration-guide/IPNIntro/#ipn-protocol-and-architecture) lists a number of checks that you should do on top of the IPN verification check. Specifically, it states the following regarding item price checking:

> Verify that the payment amount in an IPN matches the price you intend to charge. If you do not encrypt your button code, it's possible for someone to capture a button-click message and change the price it contains. If you don't check the price in an IPN against the real price, you could accept a lower payment than you want.

This is also not just one of those hypothetical security holes. WooCommerce (a fairly large WordPress e-commerce platform) came in contact with this vulnerability and patched it in their [1.6.5.1 release](https://docs.woocommerce.com/document/paypal-gateway-vulnerability-patch-wc-1-6-5-1/), stating:

> In older versions of WooCommerce, the PayPal gateway did not check the value returned from IPN. This meant that if someone changed the payment form code during checkout, they could modify the order cost sent to PayPal without affecting the order status after payment. This cost change would be obvious from the PayPal payment notifications, however could go unnoticed.

I look forward to your reply,
Mooror

**Comment by mlewis-everley** [comment]
It sounds like you are using `master`, this branch isn't supported any more (for reasons such as this one).

I would suggest using `3.0` as omnipay handles these integrations server to server and is a lot more secure.

If you want to create a pull request though, more than happy to merge it.

**Comment by mooror** [comment]
Oh, my mistake. I had no idea that your module had a branch that was more up to date. I am used to the master branch being used for the most recent version of a code base. 

In that case I suppose it would be a bit pointless to fix the "old" system, as you presumably want everyone to update to the latest version of the module. How difficult is it to upgrade?

**Comment by mooror** [comment]
Woops, didn't mean to close it quite yet.

**Comment by mlewis-everley** [comment]
Um, it shouldn't be too hard, the main difference is that you are not managing gateways via `SiteConfig`, instead you will need to configure them via config.yml (as per the silverstripe-omnipay docs: https://github.com/silverstripe/silverstripe-omnipay)

To be honest, it has been quite a while since I have looked at master branch, but as long as you setup omnipay correctly, you should be up and running pretty quickly.

**Comment by mlewis-everley** [comment]
Oh yea, you need to specify the omnipay specific gateways in your composer.json as well (for example `omnipay/dummy` for testing or `omnipay/paypal`).

If you are testing payments using the dummy gateway, make sure you read these docs: https://github.com/silverstripe/silverstripe-omnipay/blob/master/docs/en/TestPayments.md
